File: hosts/python/tabduct_host/mcp_server.py
"""Tabduct host — MCP streamable-HTTP server (the "north" edge).

HTTP server on ``127.0.0.1:<port>/mcp`` using the official MCP Python SDK's
low-level ``Server`` + ``StreamableHTTPSessionManager``. Auth + origin/host
checks (PROTOCOL.md §6) gate EVERY request in a pure-ASGI middleware that runs
BEFORE the session manager (the SDK transport consumes the request body, so the
gate must read headers first and short-circuit). Mirrors
hosts/node/src/mcp-server.js + hosts/node/src/tools.js.

The catalog is served VERBATIM from protocol/tools.schema.json — tools are built
from the JSON (name/description/inputSchema), never hand-written.
"""

from __future__ import annotations

import asyncio
import hmac
import json
import re
from typing import Any

import uvicorn
from mcp.server.lowlevel.server import Server as MCPServer
from mcp.server.streamable_http_manager import StreamableHTTPSessionManager
from mcp.types import CallToolResult, ImageContent, TextContent, Tool

from tabduct_host.bridge import Bridge, ToolError
from tabduct_host.constants import (
    BIND_HOST,
    CATALOG,
    ERR,
    MCP_PATH,
    MCP_REQUEST_MAX_BYTES,
    STOP_GRACE_S,
    allowed_hosts,
)


# --- catalog → MCP tools -------------------------------------------------------

def _type_ok(type_decl: Any, v: Any) -> bool:
    types = type_decl if isinstance(type_decl, list) else [type_decl]
    for t in types:
        if t == "string":
            if isinstance(v, str):
                return True
        elif t == "integer":
            if isinstance(v, int) and not isinstance(v, bool):
                return True
        elif t == "number":
            if isinstance(v, (int, float)) and not isinstance(v, bool):
                return True
        elif t == "boolean":
            if isinstance(v, bool):
                return True
        elif t == "array":
            if isinstance(v, list):
                return True
        elif t == "object":
            if isinstance(v, dict):
                return True
        elif t == "null":
            if v is None:
                return True
        else:
            return True  # unknown/missing type → lenient
    return False


def _validate_args(schema: dict[str, Any] | None, args: dict[str, Any]) -> str | None:
    """Minimal, dependency-free schema check (required fields, types, enums).

    additionalProperties is intentionally lenient (the hub adds routing fields).
    Returns an error string or None.
    """
    if not schema or schema.get("type") != "object":
        return None
    for req in schema.get("required", []) or []:
        if args.get(req) is None:
            return f'missing required argument "{req}"'
    for key, spec in (schema.get("properties") or {}).items():
        v = args.get(key)
        if v is None:
            continue
        if spec.get("type") and not _type_ok(spec["type"], v):
            td = spec["type"]
            td = "/".join(td) if isinstance(td, list) else td
            return f'argument "{key}" must be of type {td}'
        if spec.get("enum") and v not in spec["enum"]:
            return f'argument "{key}" must be one of: {", ".join(map(str, spec["enum"]))}'
    return None


def _to_content(tool_name: str, result: Any) -> list:
    """Convert an extension tool result into MCP content blocks.

    screenshot → image content ONLY (never dump multi-MB base64 as text).
    """
    if tool_name == "screenshot":
        data_url = (result or {}).get("dataUrl", "") if isinstance(result, dict) else ""
        m = re.match(r"^data:([^;,]+);base64,([\s\S]+)$", data_url)
        if not m:
            raise ToolError(ERR.INTERNAL, "screenshot result missing base64 dataUrl")
        mime = m.group(1)
        data = re.sub(r"\s+", "", m.group(2))
        return [ImageContent(type="image", mimeType=mime, data=data)]
    return [TextContent(type="text", text=json.dumps(result))]


def build_mcp_server(bridge: Bridge) -> MCPServer:
    """Create a low-level MCP Server with catalog tools registered on it."""
    server: MCPServer = MCPServer("tabduct", "0.0.1")

    @server.list_tools()
    async def _list_tools() -> list[Tool]:
        # Served VERBATIM from the catalog — single source of truth.
        return [
            Tool(name=t["name"], description=t.get("description"), inputSchema=t.get("inputSchema", {}))
            for t in CATALOG["tools"]
        ]

    @server.call_tool(validate_input=False)
    async def _call_tool(name: str, arguments: dict[str, Any] | None) -> CallToolResult:
        tool = next((t for t in CATALOG["tools"] if t["name"] == name), None)
        if tool is None:
            return CallToolResult(
                content=[TextContent(type="text", text=f"UNKNOWN_TOOL: Unknown tool: {name}")],
                isError=True,
            )
        args = arguments or {}
        bad = _validate_args(tool.get("inputSchema"), args)
        if bad:
            return CallToolResult(
                content=[TextContent(type="text", text=f"INVALID_ARGS: {bad}")],
                isError=True,
            )
        try:
            result = await bridge.invoke(name, args)
            content = _to_content(name, result)
            return CallToolResult(content=content, isError=False)
        except ToolError as e:
            return CallToolResult(
                content=[TextContent(type="text", text=f"{e.code}: {e.message}")],
                isError=True,
            )

    return server


# --- HTTP server ---------------------------------------------------------------

class McpHttpServer:
    """Bound on ``open``; stopped on ``close``. Loopback-only, auth-gated."""

    def __init__(self, bridge: Bridge) -> None:
        self._bridge = bridge
        self._uvicorn: uvicorn.Server | None = None
        self._serve_task: asyncio.Task[None] | None = None
        self._session_manager: StreamableHTTPSessionManager | None = None
        self._sm_ctx: Any = None  # the session_manager.run() async context
        self._token: str | None = None
        self._port: int | None = None
        self._allowed_hosts: set[str] | None = None
        self.is_running = False
        self._starting = False

    # PROTOCOL.md §6 gatekeeper. Returns None if OK, else (status, message).
    def _reject(self, scope: dict[str, Any]) -> tuple[int, str] | None:
        origin = None
        host = None
        authorization = None
        content_length = None
        for name_b, value_b in scope.get("headers", []):
            name = name_b.decode("latin-1")
            if name == "origin":
                origin = value_b.decode("latin-1")
            elif name == "host":
                host = value_b.decode("latin-1")
            elif name == "authorization":
                authorization = value_b.decode("latin-1")
            elif name == "content-length":
                content_length = value_b.decode("latin-1")
        if origin:
            return (403, "origin not permitted")  # no non-browser MCP client sends Origin
        if host not in (self._allowed_hosts or set()):
            return (403, "bad host header — use 127.0.0.1")
        m = re.match(r"^Bearer\s+(.+)$", authorization or "")
        if not m or not hmac.compare_digest(m.group(1), self._token or ""):
            return (401, "unauthorized")
        # Cap the request body (parity with Node's MCP_REQUEST_MAX_BYTES). The SDK
        # transport would otherwise read an unbounded body into memory.
        if content_length is not None:
            try:
                if int(content_length) > MCP_REQUEST_MAX_BYTES:
                    return (413, "request body too large")
            except ValueError:
                return (400, "bad content-length")
        return None

    # --- ASGI app (auth gate → path route → session manager) -------------------

    async def __call__(self, scope: dict[str, Any], receive: Any, send: Any) -> None:
        stype = scope.get("type")
        if stype == "lifespan":
            await self._lifespan(receive, send)
            return
        if stype != "http":
            return  # we only handle http (+ lifespan)
        bad = self._reject(scope)
        if bad is not None:
            await self._send_status(send, bad[0], bad[1])
            return
        path = scope.get("path", "")
        if path == MCP_PATH:
            # session_manager.handle_request is the ASGI entry that creates/reuses
            # sessions and dispatches JSON-RPC to the MCP server.
            assert self._session_manager is not None
            await self._session_manager.handle_request(scope, receive, send)
        else:
            await self._send_status(send, 404, "not found")

    async def _send_status(self, send: Any, status: int, message: str) -> None:
        body = message.encode("utf-8")
        await send(
            {
                "type": "http.response.start",
                "status": status,
                "headers": [
                    (b"content-type", b"text/plain; charset=utf-8"),
                    (b"content-length", str(len(body)).encode("ascii")),
                ],
            }
        )
        await send({"type": "http.response.body", "body": body})

    async def _lifespan(self, receive: Any, send: Any) -> None:
        while True:
            msg = await receive()
            mtype = msg.get("type")
            if mtype == "lifespan.startup":
                try:
                    assert self._session_manager is not None
                    self._sm_ctx = self._session_manager.run()
                    await self._sm_ctx.__aenter__()
                    await send({"type": "lifespan.startup.complete"})
                except Exception as e:
                    await send({"type": "lifespan.startup.failed", "message": str(e)})
            elif mtype == "lifespan.shutdown":
                try:
                    if self._sm_ctx is not None:
                        await self._sm_ctx.__aexit__(None, None, None)
                        self._sm_ctx = None
                    await send({"type": "lifespan.shutdown.complete"})
                except Exception as e:
                    await send({"type": "lifespan.shutdown.failed", "message": str(e)})
                break

    # --- lifecycle -------------------------------------------------------------

    async def start(self, port: int, token: str) -> int:
        if self.is_running or self._starting:
            raise RuntimeError("Server already running")
        self._starting = True
        self._token = token
        try:
            self._session_manager = StreamableHTTPSessionManager(
                app=build_mcp_server(self._bridge),
            )
            config = uvicorn.Config(
                app=self,  # this object is the ASGI app
                host=BIND_HOST,
                port=port,
                loop="asyncio",
                http="h11",
                lifespan="on",
                access_log=False,
                log_level="error",
                timeout_graceful_shutdown=STOP_GRACE_S,
            )
            self._uvicorn = uvicorn.Server(config)
            self._serve_task = asyncio.create_task(self._uvicorn.serve())
            # Wait until the listening socket is up so we can read the bound port.
            for _ in range(500):  # up to ~5s
                if self._uvicorn.started and self._uvicorn.servers:
                    break
                if self._serve_task.done():
                    break
                await asyncio.sleep(0.01)
            if not (self._uvicorn.started and self._uvicorn.servers):
                # Surface the failure reason if the serve task died.
                exc = self._serve_task.exception() if self._serve_task.done() else None
                raise RuntimeError(f"server failed to start: {exc}")
            sockname = self._uvicorn.servers[0].sockets[0].getsockname()
            self._port = int(sockname[1])
            self._allowed_hosts = allowed_hosts(self._port)
            self.is_running = True
            return self._port
        finally:
            self._starting = False

    async def stop(self) -> None:
        uv = self._uvicorn
        task = self._serve_task
        self._uvicorn = None
        self._serve_task = None
        if uv is not None and task is not None:
            uv.should_exit = True
            try:
                await asyncio.wait_for(task, timeout=10)
            except asyncio.TimeoutError:
                uv.force_exit = True
                try:
                    await asyncio.wait_for(task, timeout=5)
                except (asyncio.TimeoutError, Exception):
                    pass
        # Closes the listening socket (→ connections refused afterwards) and runs
        # lifespan shutdown which exits the session_manager context (reaps sessions).
        self._session_manager = None
        self._sm_ctx = None
        self._token = None
        self._port = None
        self._allowed_hosts = None
        self.is_running = False

File: hosts/dotnet/Tabduct.Host/McpServer.cs
using System.Collections.Concurrent;
using System.Net;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Server.Kestrel.Core;
using Microsoft.AspNetCore.Server.Kestrel.Transport;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using ModelContextProtocol.Protocol;
using ModelContextProtocol.Server;

namespace Tabduct.Host;

/// <summary>
/// MCP streamable-HTTP server (the "north" edge). HTTP server on
/// <c>127.0.0.1:&lt;port&gt;/mcp</c> using the official C# MCP SDK's low-level
/// request handlers + the ASP.NET Core streamable-HTTP transport. Auth + origin/host
/// checks (PROTOCOL.md §6) gate EVERY request in middleware that runs BEFORE the
/// MCP endpoint. Mirrors hosts/python/tabduct_host/mcp_server.py +
/// hosts/node/src/mcp-server.js + tools.js.
///
/// The catalog is served VERBATIM from protocol/tools.schema.json (single source
/// of truth) — tools are built from the JSON, never hand-written.
/// </summary>
internal sealed class McpServer
{
    public const string BindHost = "127.0.0.1";
    public const string McpPath = "/mcp";
    public const int DefaultPort = 12310;

    // Loaded once at process start; the JsonDocument is kept alive so the cloned
    // inputSchema JsonElements remain valid for the process lifetime.
    private static readonly JsonDocument CatalogDoc = LoadCatalog();
    public static readonly int ProtocolVersion = CatalogDoc.RootElement.GetProperty("protocolVersion").GetInt32();

    private static readonly CatalogEntry[] Catalog = BuildCatalog(CatalogDoc);
    private static readonly Tool[] McpTools = BuildMcpTools(Catalog);

    private static readonly Regex BearerRe = new("^Bearer\\s+(.+)$", RegexOptions.Compiled);
    private static readonly Regex DataUrlRe = new("^data:([^;,]+);base64,([\\s\\S]+)$", RegexOptions.Compiled);
    private static readonly Regex WhitespaceRe = new("\\s+", RegexOptions.Compiled);

    private readonly Bridge _bridge;
    private WebApplication? _app;
    private ListenOptions? _listen;
    private string? _token;
    private int _port;
    public bool IsRunning { get; private set; }
    private bool _starting;

    public McpServer(Bridge bridge) => _bridge = bridge;

    /// <summary>Bind on <paramref name="port"/> (0 = ephemeral); returns the actually-bound port.</summary>
    public async Task<int> Start(int port, string token)
    {
        if (IsRunning || _starting) throw new InvalidOperationException("Server already running");
        _starting = true;
        try
        {
            _token = token;
            _listen = null;

            var builder = WebApplication.CreateSlimBuilder();
            // CRITICAL: keep stdout pure native-messaging frames. No console/kestrel
            // logging to stdout. (We log nothing by default; nothing reaches stdout.)
            builder.Logging.ClearProviders();

            builder.WebHost.ConfigureKestrel(opts =>
            {
                // Cap the request body at the protocol limit (parity with Node's
                // MCP_REQUEST_MAX_BYTES) rather than leaning on Kestrel's ~30 MB default.
                opts.Limits.MaxRequestBodySize = 8 * 1024 * 1024;
                opts.Listen(IPAddress.Loopback, port, lo => _listen = lo);
            });

            builder.Services.AddMcpServer(o =>
                {
                    o.ServerInfo = new Implementation { Name = "tabduct", Version = "0.0.1" };
                })
                .WithListToolsHandler((_, _) => new ValueTask<ListToolsResult>(new ListToolsResult { Tools = McpTools }))
                .WithCallToolHandler((ctx, ct) => new ValueTask<CallToolResult>(CallTool(ctx.Params!, ct)))
                .WithHttpTransport();

            var app = builder.Build();

            // PROTOCOL.md §6 gatekeeper — runs BEFORE the MCP endpoint consumes the body.
            app.Use(async (http, next) =>
            {
                // 1. Any Origin header → 403 (no non-browser MCP client sends Origin).
                if (!string.IsNullOrEmpty(http.Request.Headers.Origin.ToString()))
                {
                    http.Response.StatusCode = StatusCodes.Status403Forbidden;
                    return;
                }
                // 2. Host header must be a loopback alias on our port (DNS-rebinding defense).
                var allowed = AllowedHosts(_port);
                if (!allowed.Contains(http.Request.Headers.Host.ToString()))
                {
                    http.Response.StatusCode = StatusCodes.Status403Forbidden;
                    return;
                }
                // 3. Authorization: Bearer <token> (constant-time compare).
                string? bearer = ParseBearer(http.Request.Headers.Authorization.ToString());
                if (bearer is null || !TokenEquals(bearer, _token ?? ""))
                {
                    http.Response.StatusCode = StatusCodes.Status401Unauthorized;
                    return;
                }
                await next();
            });

            app.MapMcp(McpPath);

            _app = app;
            await app.StartAsync();

            _port = _listen!.IPEndPoint!.Port;
            IsRunning = true;
            return _port;
        }
        finally
        {
            _starting = false;
        }
    }

    public async Task Stop()
    {
        var app = _app;
        _app = null;
        if (app is not null)
        {
            // StopAsync closes the listening socket → connections REFUSED afterwards,
            // and reaps the MCP sessions via the streamable-HTTP transport shutdown.
            await app.StopAsync(TimeSpan.FromSeconds(2));
            await app.DisposeAsync();
        }
        _listen = null;
        _token = null;
        _port = 0;
        IsRunning = false;
    }

    // --- catalog → MCP tools --------------------------------------------------

    private sealed record CatalogEntry(string Name, string? Description, JsonElement InputSchema);

    private static JsonDocument LoadCatalog()
    {
        string path = ResolveCatalogPath();
        return JsonDocument.Parse(File.ReadAllText(path));
    }

    private static string ResolveCatalogPath()
    {
        // 1. CWD-relative (conformance runs from the repo root).
        string cwdRel = Path.Combine(Environment.CurrentDirectory, "protocol", "tools.schema.json");
        if (File.Exists(cwdRel)) return cwdRel;

        // 2. Walk up from the assembly base dir (built dll lives several levels under the repo).
        for (var dir = new DirectoryInfo(AppContext.BaseDirectory); dir is not null; dir = dir.Parent)
        {
            string p = Path.Combine(dir.FullName, "protocol", "tools.schema.json");
            if (File.Exists(p)) return p;
        }
        // 3. Walk up from CWD as a last resort.
        for (var dir = new DirectoryInfo(Environment.CurrentDirectory); dir is not null; dir = dir.Parent)
        {
            string p = Path.Combine(dir.FullName, "protocol", "tools.schema.json");
            if (File.Exists(p)) return p;
        }
        throw new FileNotFoundException("protocol/tools.schema.json not found");
    }

    private static CatalogEntry[] BuildCatalog(JsonDocument doc)
    {
        var tools = doc.RootElement.GetProperty("tools");
        var list = new List<CatalogEntry>();
        foreach (var t in tools.EnumerateArray())
        {
            string name = t.GetProperty("name").GetString()!;
            string? description = t.TryGetProperty("description", out var d) ? d.GetString() : null;
            JsonElement schema = t.TryGetProperty("inputSchema", out var s) ? s : default;
            // Clone so the element is independent of the document's pooled buffers.
            list.Add(new CatalogEntry(name, description, schema.Clone()));
        }
        return list.ToArray();
    }

    private static Tool[] BuildMcpTools(CatalogEntry[] catalog)
        => catalog.Select(t => new Tool
        {
            Name = t.Name,
            Description = t.Description ?? "",
            InputSchema = t.InputSchema,
        }).ToArray();

    // --- tools/call -----------------------------------------------------------

    private async Task<CallToolResult> CallTool(CallToolRequestParams p, CancellationToken ct)
    {
        string name = p.Name ?? "";
        var entry = Array.Find(Catalog, t => t.Name == name);
        if (entry is null)
        {
            return Error($"UNKNOWN_TOOL: Unknown tool: {name}");
        }

        JsonObject args = ToArgsObject(p.Arguments);
        string? bad = ValidateArgs(entry.InputSchema, args);
        if (bad is not null)
        {
            return Error($"INVALID_ARGS: {bad}");
        }

        try
        {
            JsonNode? result = await _bridge.Invoke(name, args).ConfigureAwait(false);
            return ToContent(name, result);
        }
        catch (Bridge.ToolError e)
        {
            return Error($"{e.Code}: {e.Message}");
        }
    }

    private static CallToolResult Error(string text) => new()
    {
        Content = new List<ContentBlock> { new TextContentBlock { Text = text } },
        IsError = true,
    };

    private static JsonObject ToArgsObject(IDictionary<string, JsonElement>? args)
    {
        var obj = new JsonObject();
        if (args is null) return obj;
        foreach (var kv in args)
        {
            obj[kv.Key] = JsonNode.Parse(kv.Value.GetRawText());
        }
        return obj;
    }

    /// <summary>Convert an extension tool result into MCP content blocks (screenshot → image ONLY).</summary>
    private static CallToolResult ToContent(string toolName, JsonNode? result)
    {
        if (toolName == "screenshot")
        {
            string dataUrl = result?["dataUrl"]?.GetValue<string>() ?? "";
            var m = DataUrlRe.Match(dataUrl);
            if (!m.Success)
            {
                return Error("INTERNAL: screenshot result missing base64 dataUrl");
            }
            string mime = m.Groups[1].Value;
            string data = WhitespaceRe.Replace(m.Groups[2].Value, "");
            return new CallToolResult
            {
                Content = new List<ContentBlock>
                {
                    new ImageContentBlock { MimeType = mime, Data = Convert.FromBase64String(data) },
                },
            };
        }

        return new CallToolResult
        {
            Content = new List<ContentBlock>
            {
                new TextContentBlock { Text = JsonSerializer.Serialize(result) },
            },
        };
    }

    // --- minimal, dependency-free inputSchema validation (required/type/enum) --

    private static string? ValidateArgs(JsonElement schema, JsonObject args)
    {
        if (schema.ValueKind != JsonValueKind.Object) return null;
        if (!schema.TryGetProperty("type", out var typeEl) || typeEl.GetString() != "object") return null;

        if (schema.TryGetProperty("required", out var reqEl) && reqEl.ValueKind == JsonValueKind.Array)
        {
            foreach (var r in reqEl.EnumerateArray())
            {
                string? req = r.GetString();
                if (req is not null && args[req] is null)
                {
                    return $"missing required argument \"{req}\"";
                }
            }
        }

        if (schema.TryGetProperty("properties", out var propsEl) && propsEl.ValueKind == JsonValueKind.Object)
        {
            foreach (var prop in propsEl.EnumerateObject())
            {
                string key = prop.Name;
                if (!args.TryGetPropertyValue(key, out var v) || v is null) continue;
                var spec = prop.Value;

                if (spec.TryGetProperty("type", out var typeDecl) && !TypeOk(typeDecl, v))
                {
                    string td = typeDecl.ValueKind == JsonValueKind.Array
                        ? string.Join("/", typeDecl.EnumerateArray().Select(x => x.GetString()))
                        : typeDecl.GetString() ?? "";
                    return $"argument \"{key}\" must be of type {td}";
                }

                if (spec.TryGetProperty("enum", out var enumEl) && enumEl.ValueKind == JsonValueKind.Array)
                {
                    bool found = false;
                    foreach (var e in enumEl.EnumerateArray())
                    {
                        if (JsonNodeEqual(e, v)) { found = true; break; }
                    }
                    if (!found)
                    {
                        string opts = string.Join(", ", enumEl.EnumerateArray().Select(x => x.ToString()));
                        return $"argument \"{key}\" must be one of: {opts}";
                    }
                }
            }
        }

        return null;
    }

    private static bool TypeOk(JsonElement typeDecl, JsonNode v)
    {
        IEnumerable<string> types = typeDecl.ValueKind == JsonValueKind.Array
            ? typeDecl.EnumerateArray().Select(x => x.GetString() ?? "")
            : new[] { typeDecl.GetString() ?? "" };

        foreach (var t in types)
        {
            switch (t)
            {
                case "string": if (v is JsonValue jv && jv.TryGetValue<string>(out _)) return true; break;
                case "integer": if (IsInteger(v)) return true; break;
                case "number": if (IsNumber(v)) return true; break;
                case "boolean": if (v is JsonValue bv && bv.TryGetValue<bool>(out _)) return true; break;
                case "array": if (v is JsonArray) return true; break;
                case "object": if (v is JsonObject) return true; break;
                case "null": if (v is null) return true; break;
                default: return true; // unknown/missing → lenient
            }
        }
        return false;
    }

    private static bool IsInteger(JsonNode v)
    {
        if (v is not JsonValue jv) return false;
        return jv.TryGetValue(out decimal d) ? d == Math.Truncate(d) : jv.TryGetValue(out long _);
    }

    private static bool IsNumber(JsonNode v)
        => v is JsonValue jv && (jv.TryGetValue(out decimal _) || jv.TryGetValue(out double _));

    private static bool JsonNodeEqual(JsonElement el, JsonNode? node)
    {
        if (node is null) return el.ValueKind == JsonValueKind.Null;
        return JsonNode.Parse(el.GetRawText())?.ToJsonString() == node.ToJsonString();
    }

    // --- auth helpers ---------------------------------------------------------

    private static HashSet<string> AllowedHosts(int port)
        => new() { $"{BindHost}:{port}", $"localhost:{port}", $"[::1]:{port}" };

    private static string? ParseBearer(string authorization)
    {
        if (string.IsNullOrEmpty(authorization)) return null;
        var m = BearerRe.Match(authorization);
        return m.Success ? m.Groups[1].Value : null;
    }

    /// <summary>Constant-time bearer-token compare on UTF-8 bytes.</summary>
    private static bool TokenEquals(string a, string b)
    {
        byte[] ba = Encoding.UTF8.GetBytes(a);
        byte[] bb = Encoding.UTF8.GetBytes(b);
        return ba.Length == bb.Length && CryptographicOperations.FixedTimeEquals(ba, bb);
    }
}
